Automattic/vip-cli
# SQL Import: Get presigned S3 URLs from the API, Upload SQL files, & kick off an import

## Description

Includes:
* Validate the SQL file before attempting to upload it.
* Compress uncompressed files greater than a threshold
* Get a pre-signed upload URLs from the public API (which gets it from the private API)
* If the file is larger than an arbitrary size threshold, use the S3 Multipart uploading scheme:
  * Get pre-signed upload URL / request data to upload each piece of the file
  * Use the URLs to `PUT` each piece
  * Get pre-signed upload URL / request data to "Complete" the transfer
  * Use the URL to complete the transfer
* ...or if the file is smaller, just use the `PutObject` API
* Calculate the md5 checksum
* Include required info in a public API call to `StartImport`
* tests!

## Steps to Test

### Manually Test

1. Run the public API with the branch & instructions in Automattic/vip-go-api-public#965 (including the GOOP piece)
1. Check out this PR.
1. Run `npm run build`
1. Run `API_HOST=http://localhost:4000 ./dist/bin/vip-import-sql.js @1000.staging ./__fixtures__/client-file-uploader/db-dump-ipsum-67mb.sql`
    1. You should see no errors and the file should be uploaded in a single part
1. Run `API_HOST=http://localhost:4000 ./dist/bin/vip-import-sql.js @1000.staging $PATH_TO_SOME_LARGEISH_SQL_FILE` (where `$PATH_TO_SOME_LARGEISH_SQL_FILE` is any file that's larger than about 100 MB -- so it won't be smaller than 32MB when gzipped)
    1. You should see no errors and the file should be uploaded in multiple parts
1. After each SQL file is uploaded, a `StartImport` mutation should be properly initiated (Follow the test plan in each server for the expected behavior)

### Automated Tests

`npm run jest __tests__/lib/client-file-uploader.js`
